src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -111,7 +180,11 @@ private function completeDefinition($id, Definition $definition)
                 if (!$typeHint = $parameter->getClass()) {
                     // no default value? Then fail
                     if (!$parameter->isOptional()) {
-                        throw new RuntimeException(sprintf('Unable to autowire argument index %d ($%s) for the service "%s". If this is an object, give it a type-hint. Otherwise, specify this argument\'s value explicitly.', $index, $parameter->name, $id));
+                        if ($isConstructor) {
+                            throw new RuntimeException(sprintf('Unable to autowire argument index %d ($%s) for the service "%s". If this is an object, give it a type-hint. Otherwise, specify this argument\'s value explicitly.', $index, $parameter->name, $id));
```

Comment:
What about reversing the `if` condition here to make the exception message more readable?
```php
if (!$isConstructor) {
    return;
}

// ...